spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala

```diff
@@ -0,0 +1,112 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.ml
+import com.intel.analytics.bigdl.dataset.{DataSet, MiniBatch}
+import com.intel.analytics.bigdl.{Criterion, Module}
+import com.intel.analytics.bigdl.optim.Optimizer
+import com.intel.analytics.bigdl.tensor.{Tensor}
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import org.apache.spark.ml.param.shared.HasInputCols
+import org.apache.spark.ml.param.{Param, ParamMap, Params}
+import org.apache.spark.sql.{DataFrame}
+import org.apache.spark.rdd.RDD
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+/**
+ * A wrapper of Optimizer to support fit() in ML Pipelines as an Estimator
+ */
+class DLEstimator[@specialized(Float, Double) T: ClassTag]
+(override val uid: String = "DLEstimator")(implicit ev: TensorNumeric[T])
+  extends MLEstimator with HasInputCols with DLDataParams[T]{
+
+  def setInputCols(inputColName: Array[String]): this.type = set(inputCols, inputColName)
+
+  def validateParameters(): Unit = {
+    val params = this.extractParamMap()
+    require(null != params.getOrElse(modelTrain, null),
+      "DLEstimator: model for optimization must not be null")
+    require(null != params.getOrElse(batchShape, null),
+      "DLEstimator: batchShape for optimization must not be null")
+    require(null != params.getOrElse(inputCols, null),
+      "DLEstimator: inputCols must not be null")
+    require(null != params.getOrElse(criterion, null),
+      "DLEstimator: criterion must not be null")
+  }
+
+  override def process(dataFrame: DataFrame): MlTransformer = {
+
+    validateParameters()
+
+    val batches = toMiniBatch(dataFrame)
+
+    val dataset = DataSet.rdd(batches)
+
+    val optimizer = Optimizer($(modelTrain), dataset, $(criterion))
+
+    var optimizedModule = $(modelTrain)
+
+    optimizedModule = optimizer.optimize()
+
+    var classifier = new DLClassifier[T]()
+      .setInputCol("features")
+      .setOutputCol("predict")
+
+    val paramsTrans = ParamMap(
+      classifier.modelTrain -> optimizedModule,
+      classifier.batchShape -> ${batchShape})
+
+    classifier = classifier.copy(paramsTrans)
+
+    classifier
+  }
+
+  private def toMiniBatch(dataFrame: DataFrame) : RDD[MiniBatch[T]] = {
+    val inputs = $(inputCols)
+    require(inputs.length == 4, "Input columns size " +
+      s" ${inputs.length} != 4 ,which stands for feature data,feature size," +
+      s" label data and lable size respectively")
+    val data = dataFrame.select(inputs.map(input => dataFrame(input)) : _*).rdd
+    val batchs = data.map(row => {
+      val featureData = row.getAs[mutable.WrappedArray[T]](0).toArray
+      val featureSize = row.getAs[mutable.WrappedArray[Int]](1).toArray
```

Comment:
Refined, we also get rid of indexing by doing so as we only have feature data and label data in DF now